File: Week 6/2021_S3.cpp
#include <bits/stdc++.h> 
using namespace std; 

#define ll long long

//calculates the total walking time for a given position
ll calculate_time(ll pos, ll N, vector<vector<ll>> friends) {

    ll time = 0; 

    for(int i = 0; i < N; i++) {
        time += (max(abs(pos - friends[i][0]) - friends[i][2], (ll)0) * friends[i][1]); 
    }

    return time; 
}


int main() {

    ios_base::sync_with_stdio(false); 
    cin.tie(0); 

    ll N; 
    cin >> N; 

    //stores friends
    vector<vector<ll>> friends(N, vector<ll>(3)); 

    //left and right bounds for binary search
    ll left = 1000000000; 
    ll right = 0; 
    
    // update friends array and update the furthest left and right bounds
    for(int i = 0; i < N; i++) {
        cin >> friends[i][0] >> friends[i][1] >> friends[i][2]; 

        right = max(right, friends[i][0]); 
        left = min(left, friends[i][0]); 
    }

    // calculate starting mid point
    ll mid = (right + left) / 2; 
    
    // perform binary search to find smallest value
    while(true) {
        ll left_time = calculate_time(mid-1, N, friends); 
        ll right_time = calculate_time(mid+1, N, friends); 
        ll cur_time = calculate_time(mid, N, friends); 

        if(left_time < cur_time) {
            right = mid; 
            mid = (mid + left) / 2; 
        } else if(right_time < cur_time) {
            left = mid; 
            mid = (right + mid) /2; 
        } else {
            cout << cur_time; 
            break; 
        }

    }
    
    return 0; 
}

File: Week 6/2021_S3.py
numFriends = int(input())

friends = [[0, 0, 0]] * numFriends

#Called a "number line," use binary search since sorted
left = 1000000000
right = 0

for x in range(numFriends):
	values = input().split(" ")
	#Initial Pos, Walking Rate m/s, Hearing Distance 
	friends[x] = [int(values[0]), int(values[1]), int(values[2])]
	right = max(right, friends[x][0])
	left = min(left, friends[x][0])

mid = (left + right) / 2

def calculateTime(pos):
	totalTime = 0

	for x in range(numFriends):
		#Absolute of their desired pos - current pos, - hearing distance, and multiply by rate in m/s to find time taken per metre traveled
		totalTime += (max(abs(pos - friends[x][0]) - friends[x][2], 0) * friends[x][1])

	return totalTime

while True:
	leftTime = calculateTime(mid - 1)
	rightTime = calculateTime(mid + 1)
	currentTime = calculateTime(mid)

	#Compare times and directions to move along the number line
	if leftTime < currentTime:
		right = mid
		mid = (left + mid) / 2
	elif rightTime < currentTime:
		left = mid
		mid = (right + mid) / 2
	else:
		print(int(currentTime))
		break